File: DailyChallenges/CPP/LeetCode_2359.cpp
class Solution {
public:
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        if (node1 == node2) return node1;
        int n = edges.size();
        vector<int> v(n); v[node1] = 1; v[node2] = -1;
        queue<int> q; q.push(node1); q.push(node2);
        while (!q.empty()){
            set<int> s;
            int sz = q.size();
            for (int i = 0; i<sz; ++i){
                int st = q.front(); q.pop();
                int ed = edges[st]; 
                if (ed == -1) continue;
                if (v[ed] == -v[st]) 
                    s.insert(ed);
                else if (v[ed] == v[st]) 
                    continue;
                else{
                    v[ed] = v[st]; q.push(ed);
                }
            }
            if (!s.empty()) return *s.begin();
        }
        return -1;
    }
};

File: DailyChallenges/Scala/LeetCode_2359.scala
object Solution {
    def closestMeetingNode(edges: Array[Int], node1: Int, node2: Int): Int = {
        if (node1 == node2) return node1
        val n = edges.length
        val v = Array.ofDim[Int](n); v(node1) = 1; v(node2) = -1
        val q = scala.collection.mutable.Queue.empty[Int]; q.enqueue(node1); q.enqueue(node2)
        while (q.nonEmpty){
            var l: List[Int] = Nil
            val sz = q.length
            for (i <- 0 until q.length; st = q.dequeue; ed = edges(st) if (ed != -1 && v(ed) != v(st))){
                if (v(ed) == -v(st))
                    l = ed :: l
                else{
                    v(ed) = v(st)
                    q.enqueue(ed)
                }
            }
            if (l.nonEmpty) return l.sorted.head
        }
        -1
    }
}
